src/System.IO.Pipelines/Pipe.cs

```diff
@@ -11,6 +11,10 @@ namespace System.IO.Pipelines
     /// </summary>
     internal class Pipe : IPipe, IPipeReader, IPipeWriter, IReadableBufferAwaiter, IWritableBufferAwaiter
     {
+        // This sync objects protects the following state:
+        // 1. _commitHead & _commitHeadIndex
```

Comment:
I think mimicking socket input at this point is strictly necessary. So let's overlock unnecessarily but parity would be good.